rust/kernel/usb.rs

```diff
@@ -0,0 +1,421 @@
+/// ```ignore
+/// # use kernel::{usb, define_usb_id_table};
+/// #
+/// struct MyDriver;
+/// impl usb::Driver for MyDriver {
+///     // [...]
+/// #   fn probe(_dev: &mut usb::Device, _id_info: Option<&Self::IdInfo>) -> Result {
+/// #       Ok(())
+/// #   }
+/// #   define_usb_id_table! {u64, [
+/// #       (usb::DeviceId::new(0xbee7, 0xffff), None),
+/// #       (usb::DeviceId::with_class(0xbee7, 0xff1d), Some(0x40)),
+/// #   ]}
+/// }
+#[macro_export]
+macro_rules! define_usb_id_table {
+    ($data_type:ty, $($t:tt)*) => {
+        type IdInfo = $data_type;
+        const ID_TABLE: $crate::driver::IdTable<'static, $crate::usb::DeviceId, $data_type> = {
+            $crate::define_id_array!(ARRAY, $crate::usb::DeviceId, $data_type, $($t)* );
+            ARRAY.as_table()
+        };
+    };
+}
+
+/// An USB driver
+pub trait Driver {
+    /// Data stored on device interface by driver.
+    ///
+    /// Corresponds to the data set or retrieved via the kernel's
+    /// `usb_{set,get}_intfdata()` functions.
+    ///
+    /// Require that `Data` implements `PointerWrapper`. We guarantee to
+    /// never move the underlying wrapped data structure.
+    type Data: PointerWrapper + Send + Sync + driver::DeviceRemoval = ();
+
+    /// The type holding information about each device id supported by the driver.
+    /// Trait `From<NonZeroU64>` is required due to driver info being either a
+    /// flagset or a pointer to driver-specific data.
+    type IdInfo: 'static + From<NonZeroU64> = u64;
+
+    /// The table of device ids supported by the driver.
+    const ID_TABLE: driver::IdTable<'static, DeviceId, Self::IdInfo>;
+
+    /// USB driver probe.
+    ///
+    /// Called to see if the driver can manage a device interface.
+    /// Implementer should attempt to initialize the interface here.
+    fn probe(dev: &mut Device, id: Option<&Self::IdInfo>) -> Result<Self::Data>;
+
+    /// USB driver disconnect.
+    ///
+    /// Called when a device interface is removed.
+    /// Implementer should prepare the interface for removal here.
+    fn disconnect(data: &Self::Data);
+}
+
+/// An USB device.
+///
+/// # Invariants
+///
+/// The field `ptr` is non-null and valid for the lifetime of the object.
+pub struct Device {
+    ptr: *mut bindings::usb_device,
+}
+
+impl Device {
+    unsafe fn from_ptr(ptr: *mut bindings::usb_device) -> Self {
+        Self { ptr }
+    }
+}
+
+unsafe impl device::RawDevice for Device {
+    fn raw_device(&self) -> *mut bindings::device {
+        // SAFETY: By the type invariants, we know that `self.ptr` is non-null and valid.
+        unsafe { &mut (*self.ptr).dev }
+    }
+}
+
+/// An USB device interface.
+///
+/// # Invariants
+///
+/// The field `ptr` is non-null and valid for the lifetime of the object.
+pub struct Interface {
+    ptr: *mut bindings::usb_interface,
+}
+
+impl Interface {
+    /// Creates an interface from a raw pointer.
+    pub unsafe fn from_ptr(ptr: *mut bindings::usb_interface) -> Self {
```

Comment:
Preferably we'd like to avoid `unsafe` functions in the public interface, so that drivers can be written with as little unsafe rust as possible.

Reply:
I feel this to be exposed as a necessity in some cases, but if `From` and `Into` could be used instead then this can be privatized.

Reply:
I'm thinking about the case when we have to obtain a raw pointer to `bindings::usb_interface` from something like `bindings::hid_device` and then use said pointer with the `usb::Interface` wrapper, `from_ptr` is for now with the PR but wouldn't be better to have another method from a wrapper, `hid::Device` for example, or its `Into` trait implementation. If I go by that criterion then it's not necessary to expose `from_ptr`. What do you think?